### ISSUE TITLE
joblib.parallel_backend doesn't return when used with dask-mpi and sklearn

### ISSUE THREAD

**Issue by lebedov** [dask/dask-ml#583]
Although the following seems to run to completion when run with `mpiexec -np 4 script.py`, the context manager doesn't seem to return:

```python3
from dask_mpi import initialize
initialize()

from distributed import Client
client = Client()

import joblib
import numpy as np
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC

digits = load_digits()

param_space = {
    'C': np.logspace(-3, 3, 5),
    'gamma': np.logspace(-4, 4, 3),
}

model = SVC(kernel='rbf')
search = GridSearchCV(model, param_space, cv=3, verbose=10)

with joblib.parallel_backend('dask'):
    search.fit(digits.data, digits.target)
print('done') # this is never printed
```

Any thoughts why? I'm using python 3.7.3, scikit-learn 0.21.3, joblib 0.14.0, dask 2.7.0, and dask-mpi 2.0.0 with openmpi 4.0.2 on Ubuntu 18.04.

The motivation behind the above, incidentally, is to examine the feasibility of using dask-mpi to speed up cross validation of models that conform to the sklearn estimator interface on clusters of machines.

**Comment by lebedov** [comment]
Yes - grid_search.fit() returns but the context manager doesn't seem to close.

**Comment by lebedov** [comment]
When I use dask_ml.model_selection.GridSearchCV instead of sklearn.model_selection.GridSearchCV, the context manager closes normally. Something still seems to be going on in joblib, though, as trying to run something like

```
joblib.Parallel()(joblib.delayed(lambda x: x+1)(np.random.rand(12, 12)) for i in range(4))
```
within the context manager doesn't return (but does for smaller matrices).

**Comment by TomAugspurger** [comment]
> When I use dask_ml.model_selection.GridSearchCV instead of sklearn.model_selection.GridSearchCV, the context manager closes normally.

Haven't looked into the full issue yet, but just as an FYI, `dask_ml.model_selection.GridSearchCV` doesn't use joblib at all.
